### PR TITLE
fix(progress): clear reflowed frame after resize

### PR DESCRIPTION
**Pull request by jdx**
## Summary

- redraw stable-size progress updates with the portable cursor-up/clear path
- clear the visible viewport on the first terminal-size change, suppress rendering while dimensions keep moving, and write one fresh layout after 100 ms of stability
- suppress frames that cannot fit in a cramped viewport, then recover with one clean frame when space returns
- avoid a duplicate final redraw when `stop()` runs after background completion
- always restore cursor visibility when stopping an active UI
- keep output caching write-aware
- add PTY coverage, a rapid-resize tmux terminal-state regression, and a 20-second interactive resize demo

## Root cause

Terminal reflow can move part of an existing multi-row progress frame into scrollback before clx observes the new dimensions. Cursor movement and saved-cursor sequences cannot reliably recover that content across terminals; a real-terminal recording also showed CSI save/restore being treated as unsupported, which made ordinary refreshes accumulate.

Clearing and immediately rendering on every reported dimension was still racy during a continuous drag: the terminal could reflow the replacement frame again before dimensions settled. The portable behavior is now explicit:

- stable-size refreshes use cursor-up plus clear-to-end
- the first resize event clears the viewport and leaves it blank
- continued changes restart a 100 ms settle timer
- one final synchronized clear and redraw runs after the size remains stable

This trades visible shell history in the current viewport—and a brief blank progress area while actively dragging—for deterministic recovery without duplicated frames.

## Manual testing

```sh
cargo run --example resize
```

While it runs:

1. Leave the terminal at a stable size and confirm the spinner updates without accumulating rows.
2. Continuously drag the terminal narrower and wider. The viewport should stay blank during active movement, then show exactly one clean layout shortly after you release it.
3. Shrink it until the progress tree cannot fit. It should remain blank until it can show a complete frame.
4. Expand it again and confirm exactly one progress tree appears.
5. Let the demo finish and confirm the cursor is visible below the final output.

An automated real-terminal check is also available when tmux is installed:

```sh
CLX_TMUX_BIN="$(command -v tmux)" cargo test --test resize \
  tmux_resize_keeps_one_copy_of_each_progress_row -- --nocapture
```

The tmux test rapidly changes width in 20 ms increments in both directions, waits for settling, and verifies that each progress row appears exactly once.

## Validation

- `mise run ci`
- `cargo test --test resize -- --nocapture`
- `cargo test --test synchronized_output -- --nocapture`
- rapid tmux resize test from 120 to 40 columns and back

*This PR was generated by an AI coding assistant.*

---

> [!NOTE]
> **Medium Risk**
> Changes core TTY redraw, resize, and stop paths for all UI-mode progress; behavior trades brief blank screens during resize for correctness, with good test coverage but terminal-specific edge cases remain possible.
> 
> **Overview**
> Fixes **terminal resize and cramped-viewport** bugs where multi-row progress frames could reflow into scrollback, duplicate rows, or leave a hidden cursor.
> 
> **Resize while jobs are running:** the first size change **clears the viewport** and **defers redraws** until dimensions stay stable for **100ms**, then performs one **full-screen clear** and a single fresh layout. Stable-size updates still use **cursor-up + clear-to-end**.
> 
> **Cramped terminals:** when the rendered frame (or prior frame) would **fill the viewport**, output is **suppressed** until a complete frame fits again, with a one-time clear on entry.
> 
> **Lifecycle fixes:** `LAST_OUTPUT` updates only when a frame is **actually written**; `stop()` avoids a **duplicate final redraw** after the background thread already finished; **`finish_frame()`** restores **cursor visibility** after completion or cramped suppression.
> 
> Adds **`examples/resize`**, **PTY/tmux integration tests**, and unit tests for resize state, height calculation, and deferred caching.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 78ab70879fdf91d007c46f413ebec43642e253e1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **New Features**
  * Added terminal resize support for progress displays, including improved redraw behavior across changing terminal sizes.
  * Added a resize progress example demonstrating a long-running, dynamically updating display.

* **Bug Fixes**
  * Prevented duplicate redraws after progress completion.
  * Improved cursor visibility, screen clearing, and output stability during constrained or rapidly changing terminal viewports.

* **Tests**
  * Added coverage for PTY and tmux resizing, terminal reflow, cramped viewports, and synchronized output behavior.

### CODE REVIEW

File: src/progress/render.rs [hunk 1]
```diff
@@ -103,14 +103,31 @@ pub(crate) fn process_flex_output(output: &str) -> String {
...
+    let previous_height = rendered_height(&last_frame, term_width);
+    // Cursor-up is clamped at the top of the viewport. If a resize makes the
+    // viewport shorter than the existing frame, attempting a redraw cannot
+    // reach and erase the whole frame; each refresh would instead push another
+    // partial copy into scrollback. Leave the existing frame alone until the
+    // viewport is tall enough to address it again. Terminal states still get
+    // one best-effort final render so completed status is not lost.
+    let any_running = jobs.iter().any(|job| job.is_running());
+    if any_running && previous_height >= term_height {
+        return Ok(());
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/render.rs]: The equal-height guard is intentional. `Term::write_line` leaves the cursor on the row after the frame; when the frame's physical height equals the viewport height, its first row has already moved above the viewport. ANSI cursor-up is clamped at the top margin, so it cannot reach and clear that row. Changing this to `>` would reintroduce the partial-clear/scrollback duplication at the exact boundary. I added an explicit PTY regression case for `rows == previous_height` and kept the guard as `>=`.

*This comment was generated by an AI coding assistant.*

File: src/progress/state.rs [hunk 2]
```diff
@@ -339,10 +340,9 @@ fn start() {
 pub fn stop() {
     STOPPING.store(true, Ordering::Relaxed);
     let _ = refresh_once();
+    let _ = finish_frame();
     clear_osc_progress();
     *STARTED.lock().unwrap() = false;
-    // Reset LINES to prevent subsequent stop_clear() from clearing user output
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/state.rs:344]: Fixed in 1160c45. `stop()` now takes the refresh lock and only flushes a final frame while the progress loop is still marked started, so a background-completed frame is not appended again. The synchronized-output PTY test now waits for automatic completion, calls `stop()`, and asserts that no progress content is emitted between markers.

*This comment was generated by an AI coding assistant.*

File: src/progress/state.rs [hunk 3]
```diff
@@ -339,10 +344,10 @@ fn start() {
 pub fn stop() {
     STOPPING.store(true, Ordering::Relaxed);
     let _ = refresh_once();
+    let _ = finish_frame();
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/state.rs:355]: Fixed in 1160c45. `finish_frame()` now restores cursor visibility for every active UI-mode finish, including when cramped-viewport suppression has reset `LINES` to zero. A PTY regression shrinks the viewport until suppression clears/hides the display, stops while the job is still running, and verifies a show-cursor sequence follows the reset.

*This comment was generated by an AI coding assistant.*

File: src/progress/state.rs [hunk 4]
```diff
@@ -338,11 +343,11 @@ fn start() {
 /// Stops the progress display and renders the final state.
 pub fn stop() {
     STOPPING.store(true, Ordering::Relaxed);
-    let _ = refresh_once();
+    let _ = refresh_once_if_started();
+    let _ = finish_frame();
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/state.rs:355]: Fixed in 6bd7e74. `stop()` now acquires `REFRESH_LOCK` before setting `STOPPING`, checks `STARTED`, and performs the final render while still holding that lock. The background refresh loop therefore cannot clear `STARTED` between the stop signal and the final-render decision. A naturally completed loop still has `STARTED == false`, preserving the no-duplicate behavior.

*This comment was generated by an AI coding assistant.*

File: src/progress/state.rs [hunk 5]
```diff
@@ -337,22 +342,32 @@ fn start() {
...
+    {
+        let _ = refresh_once_locked();
+    }
+    drop(refresh_guard);
+    let _ = finish_frame();
     clear_osc_progress();
     *STARTED.lock().unwrap() = false;
-    // Reset LINES to prevent subsequent stop_clear() from clearing user output
-    *LINES.lock().unwrap() = 0;
+    CRAMPED_VIEWPORT.store(false, Ordering::Relaxed);
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/state.rs:358]: Fixed in a9572b4. Both `stop()` and `stop_clear()` now reset `TERMINAL_RESIZE_STATE` while holding `REFRESH_LOCK`, so no pending size or settle deadline can leak into a later progress session.

*This comment was generated by an AI coding assistant.*

File: src/progress/render.rs [hunk 6]
```diff
@@ -100,42 +158,106 @@ pub(crate) fn process_flex_output(output: &str) -> String {
...
+    let term_height = term_height as usize;
+    let term_width = term_width as usize;
+    let output_height = rendered_height(output, term_width);
+    let previous_height = rendered_height(&previous_output, term_width);
+    // Once either the old or replacement frame fills the viewport, resizing
+    // can push the anchored origin into scrollback before clx receives
+    // SIGWINCH. Reset the visible viewport once, then suppress output until a
+    // complete frame fits again. Terminal states still get a best-effort final
+    // render.
+    if any_running && output_height.max(previous_height) >= term_height {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/render.rs]: Fixed in 6a138af. The cramped-viewport guard now considers the cached previous frame height only when `LINES > 0`, meaning that frame is still visible. After a resize clear or deferred draw leaves the viewport blank, a fitting current frame can render even if `LAST_OUTPUT` still contains an older tall frame. I also added a regression test covering hidden cached, visible cached, and currently oversized frames.

Validation: `mise run ci`, `cargo test --test resize -- --nocapture`, and `cargo test --test synchronized_output -- --nocapture`.

*This comment was generated by an AI coding assistant.*

File: src/progress/render.rs [hunk 7]
```diff
@@ -162,22 +295,24 @@ pub fn refresh() -> Result<bool> {
...
     }
-    *last_output = final_output.clone();
     drop(last_output);
 
-    write_frame(&final_output, &frame.jobs)?;
+    let written = write_frame(&final_output, &frame.jobs)?;
+    cache_written_output(&mut LAST_OUTPUT.lock().unwrap(), &final_output, written);
 
     if !any_running && !any_running_check() {
+        super::state::finish_frame()?;
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/progress/render.rs:315]: Fixed in 78ab708. `refresh_once_locked()` now checks `STARTED` after acquiring `REFRESH_LOCK`. If the background loop won the race, rendered the terminal frame, called `finish_frame()`, and marked the session stopped, the synchronous terminal-status refresh returns without writing a second copy. If the synchronous path wins the lock first, it still renders normally. Added a regression test that verifies a stopped-session synchronous refresh leaves the cached visible final frame untouched.

Validation: `mise run ci`, `cargo test synchronous_refresh_skips_after_background_stops`, `cargo test --test synchronized_output -- --nocapture`, and `cargo test --test resize -- --nocapture`.

*This comment was generated by an AI coding assistant.*
